packages/zoe/src/contracts/autoswap.js

```diff
@@ -43,203 +38,168 @@ export const makeContract = harden(zcf => {
+          liqTokenSupply += liquidityExtentOut;
+
+          const add = (key, obj1, obj2) =>
+            amountMaths[key].add(obj1[key], obj2[key]);
+
+          const newPoolAmounts = harden({
+            TokenA: add('TokenA', userAllocation, poolAllocation),
+            TokenB: add('TokenB', userAllocation, poolAllocation),
+            Liquidity: poolAllocation.Liquidity,
+          });
+
+          const newUserAmounts = harden({
+            TokenA: amountMaths.TokenA.getEmpty(),
+            TokenB: amountMaths.TokenB.getEmpty(),
+            Liquidity: liquidityAmountOut,
+          });
+
+          const newTempLiqAmounts = harden({
+            TokenA: amountMaths.TokenA.getEmpty(),
+            TokenB: amountMaths.TokenB.getEmpty(),
+            Liquidity: amountMaths.Liquidity.getEmpty(),
+          });
+
+          zcf.reallocate(
+            harden([offerHandle, poolHandle, tempLiqHandle]),
+            harden([newUserAmounts, newPoolAmounts, newTempLiqAmounts]),
+          );
+          zcf.complete(harden([offerHandle, tempLiqHandle]));
+          return 'Added liquidity.';
+        };
+
+        const tempLiqInvite = makeInvite(
+          injectLiquidity,
+          { seatDesc: 'autoSwap internal' },
+          { give: { Liquidity: null } },
+        );
+
+        const zoeService = zcf.getZoeService();
+        return zoeService
+          .offer(
+            tempLiqInvite,
+            proposal,
+            harden({ Liquidity: liquidityPaymentP }),
+          )
+          .then(({ outcome: outcomeP }) => outcomeP);
+      };
+
+      const removeLiquidity = offerHandle => {
+        const userAllocation = zcf.getCurrentAllocation(offerHandle);
+        const liquidityExtentIn = userAllocation.Liquidity.extent;
+
+        const poolAllocation = getPoolAllocation();
+
+        const newUserAmounts = calcAmountsToRemove(
+          harden({
+            liqTokenSupply,
+            poolAllocation,
+            liquidityExtentIn,
+          }),
+        );
+
+        const newPoolAmounts = harden({
+          TokenA: amountMaths.TokenA.subtract(
+            poolAllocation.TokenA,
+            newUserAmounts.TokenA,
+          ),
+          TokenB: amountMaths.TokenB.subtract(
+            poolAllocation.TokenB,
+            newUserAmounts.TokenB,
+          ),
+          Liquidity: amountMaths.Liquidity.add(
+            poolAllocation.Liquidity,
+            amountMaths.Liquidity.make(liquidityExtentIn),
+          ),
         });
-        return invite;
+
+        liqTokenSupply -= liquidityExtentIn;
+
+        zcf.reallocate(
+          harden([offerHandle, poolHandle]),
+          harden([newUserAmounts, newPoolAmounts]),
+        );
+        zcf.complete(harden([offerHandle]));
+        return 'Liquidity successfully removed.';
       };
 
       return harden({
-        invite: makeInvite(),
+        invite: makeInvite(() => {}), // Just used to extract instanceHandle
```

Comment:
Why not return an `addLiquidity` invite (result of `makeAddLiquidityInvite`) by default?

Reply:
Big improvement, thanks! Done.